File: Algorithms/src/2196. Create Binary Tree From Descriptions/Code/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[][]} descriptions
 * @return {TreeNode}
 */
var createBinaryTree = function (descriptions) {
  // Stores value -> TreeNode mapping
  const nodes = new Map();

  // Stores all child values
  const children = new Set();

  for (const [parent, child, isLeft] of descriptions) {
    // Create parent node if needed
    if (!nodes.has(parent)) nodes.set(parent, new TreeNode(parent));

    // Create child node if needed
    if (!nodes.has(child)) nodes.set(child, new TreeNode(child));

    // Connect child to parent
    if (isLeft === 1) nodes.get(parent).left = nodes.get(child);
    else nodes.get(parent).right = nodes.get(child);

    // Mark child
    children.add(child);
  }

  // Root never appears as a child
  for (const [value, node] of nodes) {
    if (!children.has(value)) return node;
  }

  return null;
};

File: Algorithms/src/2196. Create Binary Tree From Descriptions/Code/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        
        # Stores value -> TreeNode mapping
        nodes = {}
        
        # Stores all child values
        children = set()
        
        for parent, child, isLeft in descriptions:
            
            # Create parent node if needed
            if parent not in nodes:
                nodes[parent] = TreeNode(parent)
            
            # Create child node if needed
            if child not in nodes:
                nodes[child] = TreeNode(child)
            
            # Connect child to correct side
            if isLeft:
                nodes[parent].left = nodes[child]
            else:
                nodes[parent].right = nodes[child]
            
            # Mark child
            children.add(child)
        
        # Root never appears as a child
        for value, node in nodes.items():
            if value not in children:
                return node